### PR TITLE
#1226 Celery Retry Exponential Backoff

### PR DESCRIPTION
**Pull request by k-macmillan**
# Description

Modified celery retry policy across all the tasks listed in the issue ticket #1226 . Since we're using jitter we have to plan for an average number of retries to hit our desired retry time. We want all of these to retry for 24 hours. Jitter retries in the interval `[1, max backoff time]`. If we assume uniform distribution (which is [what they use](https://docs.python.org/3/library/random.html)) we can calculate how many retries we would need given our max backoff time. 

We have no way to guarantee a minimum retry time if we use jitter, but we should use jitter because it is best for us, our clients, and the services we use. Given it operates on a uniform distribution we can calculate how many retries we would need to average 24 hours of retries, and that's what has been done here. All of these methods should retry, on average, for 24 hours.

#1226 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

This will be tested by forcing the email path to retry over the weekend (screenshot in the Issue).

All unit tests pass. Userflows passes. QA repo passes all tests.

## Checklist

- [x] Appropriate labels added to this pull request (i.e. "enhancement", "dependencies", etc.)
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: app/celery/contact_information_tasks.py [hunk 1]
```diff
@@ -31,21 +35,12 @@ def lookup_contact_info(self, notification_id):
...
-                'Notification has been updated to technical-failure'
-            )
-
-            update_notification_status_by_id(
-                notification_id, NOTIFICATION_TECHNICAL_FAILURE, status_reason=e.failure_reason
-            )
-            raise NotificationTechnicalFailureException(message) from e
-
+        if can_retry(self.request.retries, self.max_retries):
+            current_app.logger.warning("Unable to get contact info for notificaiton id: %s, retrying", notification_id)
```
Comment (cris-oddball) [inline app/celery/contact_information_tasks.py:39]: spelling: change 'notificaiton' to 'notification'

File: app/celery/provider_tasks.py [hunk 2]
```diff
@@ -46,24 +51,16 @@ def deliver_sms(self, notification_id, sms_sender_id=None):
...
-                NOTIFICATION_TECHNICAL_FAILURE,
-                status_reason="Retries exceeded"
-            )
-            raise NotificationTechnicalFailureException(message)
+    except Exception as e:
+        current_app.logger.exception(
+            "SMS delivery for notification id: %s failed", notification_id
+        )
+        if can_retry(self.request.retries, self.max_retries):
+            current_app.logger.warning("Unable to send sms for notificaiton id: %s, retrying", notification_id)
```
Comment (cris-oddball) [inline app/celery/provider_tasks.py:59]: change 'notificaiton' to 'notification'

File: app/celery/provider_tasks.py [hunk 3]
```diff
@@ -163,24 +156,18 @@ def deliver_email(self, notification_id: str, sms_sender_id=None):
...
-            message = "RETRY FAILED: Max retries reached. " \
-                      "The task send_email_to_provider failed for notification {}. " \
-                      "Notification has been updated to technical-failure".format(notification_id)
-            update_notification_status_by_id(
-                notification_id,
-                NOTIFICATION_TECHNICAL_FAILURE,
-                status_reason="Retries exceeded"
-            )
-            raise NotificationTechnicalFailureException(message)
+                current_app.logger.warning("Unable to send email for notificaiton id: %s, retrying", notification_id)
```
Comment (cris-oddball) [inline app/celery/provider_tasks.py:169]: change 'notificaiton' to 'notification'

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: app/celery/contact_information_tasks.py (hunk 1 above)
Comment (k-macmillan) [reply app/celery/contact_information_tasks.py]: Fixed

File: app/celery/provider_tasks.py (hunk 2 above)
Comment (k-macmillan) [reply app/celery/provider_tasks.py]: Fixed

File: app/celery/provider_tasks.py (hunk 3 above)
Comment (k-macmillan) [reply app/celery/provider_tasks.py]: Fixed

File: app/celery/lookup_recipient_communication_permissions_task.py [hunk 4]
```diff
@@ -85,9 +86,10 @@ def recipient_has_given_permission(
             identifier, communication_item.va_profile_item_id, notification_id, notification_type
         )
     except VAProfileRetryableException as e:
+        current_app.logger.warning('Encountered VAProfileRetryableException for notification: %s', notification_id)
         current_app.logger.exception(e)
         try:
-            task.retry(queue=QueueNames.RETRY)
+            raise AutoRetryException(f'Found VAProfileRetryableException, autoretrying...')
```
Comment (EvanParish) [inline app/celery/lookup_recipient_communication_permissions_task.py:92]: f-string isn't necessary here

File: app/celery/lookup_va_profile_id_task.py [hunk 5]
```diff
@@ -36,18 +40,13 @@ def lookup_va_profile_id(self, notification_id):
...
-                      "Notification has been updated to technical-failure"
-
-            notifications_dao.update_notification_status_by_id(
-                notification_id, NOTIFICATION_TECHNICAL_FAILURE, status_reason=e.failure_reason
-            )
-            raise NotificationTechnicalFailureException(message) from e
+        if can_retry(self.request.retries, self.max_retries):
+            current_app.logger.warning("Unable to lookup VA Profile ID for notification id: %s, retrying",
+                                       notification_id)
+            raise AutoRetryException(f'Found MpiRetryableException, autoretrying...', e, e.args)
```
Comment (EvanParish) [inline app/celery/lookup_va_profile_id_task.py:46]: f-string isn't necessary here

File: app/celery/lookup_recipient_communication_permissions_task.py (hunk 4 above)
Comment (k-macmillan) [reply app/celery/lookup_recipient_communication_permissions_task.py]: Fixed.

File: app/celery/lookup_va_profile_id_task.py (hunk 5 above)
Comment (k-macmillan) [reply app/celery/lookup_va_profile_id_task.py]: Fixed.

File: ci/Dockerfile.test [hunk 6]
```diff
@@ -15,3 +15,4 @@ WORKDIR /app
 RUN rm -rf /temp
 USER vanotify
 CMD ["sh", "-c", "make test"]
+#CMD ["sh", "-c", "tail -f setup.cfg"]
```
Comment (EvanParish) [inline ci/Dockerfile.test:18]: Should we just remove this, or is there a reason to keep it?

File: app/celery/process_delivery_status_result_tasks.py [hunk 7]
```diff
@@ -231,15 +220,15 @@ def _get_notification_platform_status(self, provider: any, body: str, sqs_messag
...
+        raise AutoRetryException(f'Found {type(e).__name__}, autoretrying...')
 
     current_app.logger.info("retrieved delivery status: %s", notification_platform_status)
 
     # notification_platform_status cannot be None
     if notification_platform_status is None:
         current_app.logger.error("Notification Platform Status cannot be None")
         current_app.logger.debug(body)
-        self.retry(queue=QueueNames.RETRY)
+        raise AutoRetryException(f'Found no notification_platform_status, autoretrying...')
```
Comment (cris-oddball) [inline app/celery/process_delivery_status_result_tasks.py:231]: should this be an f-string?

**Review by cris-oddball — CHANGES_REQUESTED** [review]

File: app/celery/common.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+from app.dao.notifications_dao import update_notification_status_by_id
+from app.models import NOTIFICATION_TECHNICAL_FAILURE
+
+
+RETRIES_EXCEEDED = "Retries exceeded"
+
+
+def can_retry(retries: int, max_retries: int):
```
[the remark this answers is not in the document]
Comment (k-macmillan) [reply app/celery/common.py:8]: We do, good catch.

File: app/celery/process_delivery_status_result_tasks.py (hunk 7 above)
Comment (k-macmillan) [reply app/celery/process_delivery_status_result_tasks.py:231]: Darn, I went back and double checked after Evan called it out and caught another one but missed this one. Thanks!

**Review by EvanParish — APPROVED** [review]

**Review by cris-oddball — APPROVED** [review]

**Pull request merged**